keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -405,9 +408,70 @@ would implement restrictions based on the namespace.
 +       // +optional
 +       QoSResources QoSResourceQuota
  }
+```
+
+#### Scheduler improvements
+
+The first implementation phase only adds basic filtering of nodes based on
+QoS-class resources and node scoring is not altered. However, the relevant
+scheduler plugins (e.g. NodeResourcesFit and NodeResourcesBalancedAllocation)
+could be extended to do scoring based on the capacity, availability and usage
+of QoS-class resources on the nodes.
+
+#### Kubelet-initiated pod eviction
 
+QoS-class resources available on a node are dynamic in the sense that they may
+change over the lifetime of the node. E.g. re-configuration of the container
+runtime may make new types of QoS-class resources available, properties of
+existing resources may changes (e.g. the set of available classes) or some
+resources might be removed completely. It might be desirable that kubelet could
+evicts running pod that request QoS-class resources that are no more available
+on the node. This should be relatively straightforward to implement as kubelet
+knows what QoS-class resources are available on the node and also monitors all
+running pods.
 
+#### Default and limits
+
+LimitRanges is a way to specify per-container resource constraints. They might
+have limited applicability also in the context of QoS-class resources. In
+particular, LimitRanges could be used for specifying defaults for QoS-class
+resources. An additional possible usage would be to set per-pod limits on the
+usage of container-level QoS-class resources.
+
+```diff
+ // LimitRangeItem defines a min/max usage limit for any resource that matches on kind
+ type LimitRangeItem struct {
+@@ -5167,6 +5167,28 @@ type LimitRangeItem struct {
+        // MaxLimitRequestRatio represents the max burst value for the named resource
+        // +optional
+        MaxLimitRequestRatio ResourceList
++       //  QoSResources specifies the limits for QoS resources.
++       QoSResources []LimitQoSResource
++}
++
++// LimitQoSResource specifies limits of one QoS resources type.
++type LimitQoSResource struct {
++       // Name of the resource.
++       Name QoSResourceName
++       // Default specifies the default class to be assigned.
++       // +optional
++       Default string
++       // Max usage of classes
++       // +optional
++       Max []QoSResourceClassLimit
++}
++
++// QoSResourceClassLimit specifies a limit for one class of a QoS resource.
++type QoSResourceClassLimit struct {
++       // Name of the class.
++       Name string
++       // Capacity is the limit for usage of the class.
++       Capacity int64
+ }
 ```
+
+Just using LimitRanges for specifying defaults could simplify the API.
```

Comment:
This is a nice feature (per-NS defaults) but doesn't solve what happens when there isn't a default defined in a particular NS or for a particular resource.  So at best, this is a default override.

Reply:
Ha, totally agree. My wording here was twisted. What I tried to state is "Not supporting Max (i.e. only supporting Default) in LimitRanges could simplify the API". Updated the KEP